paf-mvp-operator-express/src/operator-node.ts

```diff
@@ -76,443 +73,536 @@ export interface OperatorNodeConfig extends Config {
+    if (input.body.identifiers !== undefined) {
+      setCookie(res, Cookies.identifiers, JSON.stringify(input.body.identifiers), getOperatorExpiration(), {
+        domain: this.topLevelDomain,
+      });
+    }
+    if (input.body.preferences !== undefined) {
+      setCookie(res, Cookies.preferences, JSON.stringify(input.body.preferences), getOperatorExpiration(), {
+        domain: this.topLevelDomain,
+      });
+    }
+  }
 
-      return getIdsPrefsResponseBuilder.buildResponse(sender, { identifiers, preferences });
-    };
+  private extractRequestAndContextFromHttp = <
+    TopLevelRequestType,
+    TopLevelRequestRedirectType extends { returnUrl: ReturnUrl; request: TopLevelRequestType }
+  >(
+    topLevelRequest: TopLevelRequestType | TopLevelRequestRedirectType,
+    req: Request
+  ) => {
+    // Extract request from Redirect request, if needed
+    let request: TopLevelRequestType;
+    let context: RestContext | RedirectContext;
+    if (
+      (topLevelRequest as TopLevelRequestRedirectType).returnUrl &&
+      (topLevelRequest as TopLevelRequestRedirectType).request
+    ) {
+      request = (topLevelRequest as TopLevelRequestRedirectType).request;
+      context = {
+        returnUrl: (topLevelRequest as TopLevelRequestRedirectType).returnUrl,
+        referer: req.header('referer'),
+      };
+    } else {
+      request = topLevelRequest as TopLevelRequestType;
+      context = { origin: req.header('origin') };
+    }
+
+    return { request, context };
+  };
+
+  private async getReadResponse(topLevelRequest: GetIdsPrefsRequest | RedirectGetIdsPrefsRequest, req: Request) {
+    const { request, context } = this.extractRequestAndContextFromHttp<GetIdsPrefsRequest, RedirectGetIdsPrefsRequest>(
+      topLevelRequest,
+      req
+    );
 
-    const getWriteResponse = async (
-      topLevelRequest: PostIdsPrefsRequest | RedirectPostIdsPrefsRequest,
-      req: Request,
-      res: Response
-    ) => {
-      const { request, context } = extractRequestAndContextFromHttp<PostIdsPrefsRequest, RedirectPostIdsPrefsRequest>(
-        topLevelRequest,
-        req
-      );
-      const sender = request.sender;
+    const sender = request.sender;
 
-      if (!allowedHosts[sender]?.includes(Permission.WRITE)) {
-        throw `Domain not allowed to write data: ${sender}`;
-      }
+    if (!this.allowedHosts[sender]?.includes(Permission.READ)) {
+      throw `Domain not allowed to read data: ${sender}`;
+    }
 
-      // Verify message
-      if (
-        !(await postIdsPrefsRequestVerifier.verifySignatureAndContent(
-          { request, context },
-          sender, // sender will always be ok
-          operatorHost // but operator needs to be verified
-        ))
-      ) {
-        // TODO [errors] finer error feedback
-        throw 'Write request verification failed';
-      }
+    if (
+      !(await this.getIdsPrefsRequestVerifier.verifySignatureAndContent(
```

Comment:
Use intermediary variable for a better readability.
Also, would it make sense to have the `host` injected in the instantiation instead?

Reply:
The `RequestVerifier` class has a generic method that takes a "sender" and "receiver" parameter, this is why the host has to be passed.
We could create a specific class for _operator_ verifiers, but I don't think this would be worth it.

Reply:
What about the intermediary variable, please?

I had a look `RequestVerifier`. It seems to be an over-engineered class/file for a simple case. So if you don't want to simplify `RequestVerifier` in this PR (which I can perfectly understand), leave this piece of code like that.

Reply:
I personally think an intermediary variable does not improve readability and I believe this is a matter of taste.
Regarding `RequestVerifier`, indeed I don't think this is relevant to refactor it in this PR, which was focused on one thing: refactor the operator and node classes, while limiting other refactorings (no "opportunistic changes").
The purpose of `RequestVerifier` is to provide a type-safe verification of messages (timestamp, sender, receiver) and signatures (each message type having its own definition). I don't consider this such a simple case.
If you think about improvements, can you please elaborate? Or even propose a PR with refactoring?
Thanks

Reply:
To explain a little bit more my point of view on intermediary variables: would you also create one for
```
if (!this.allowedHosts[sender]?.includes(Permission.READ)) {
``` 
or
```
if (!identifiers.some((i: Identifier) => i.type === 'paf_browser_id')) {
```

?
If not, then what would be the trigger to create one? To me, a simple method call does not require a variable. If it was a combination of multiple method calls with `&&` or `||` I would probably create variables.

Here, I understand some people would prefer a variable but to me it's just adding indirection.

Reply:
Concerning the Verifiers, here is a suggestion that can be compiled but that should be refined. It is enough to get the idea: https://github.com/prebid/paf-mvp-implementation/pull/225

Briefly, I rely on verifying and manipulating smart objects. No more JSON-based object that we cast to a generated model. It implies enforcement of the request verifications and at the same time some decoupling for different kind of verification which seems really positive. With a bit of rework, the `deserializexxxx` methods could be express handlers themselves. Then the real handlers are focused on business logic.


Concerning the intermediary variable, I understand your point of view. However, it is not a common practice that we see in the dev communities. If you have Clean Code in your hand, it matches the code smell `G19`. I copy/past part of it here. You will find an example in the book. 

> G19: Use Explanatory Variables
> Kent Beck wrote about this in his great book Smalltalk Best Practice Patterns and again more recently in his equally great book Implementation Patterns. One of the more powerful ways to make a program readable is to break the calculations up into intermediate values that are held in variables with meaningful names.

Given your examples, I personally give my way of thinking until there. A value should be assigned to a variable and not directly put in an `if` if: 
* if we have asynchronous stuff, 
* if the function call goes on many lines,
* there is more than one or two operations (e.g `this.allowedHosts[sender]?.includes(Permission.READ)` has 2)

Reply:
[intermediary variables] I dislike the fact that it introduces a lot of new constants in the scope of the methods (adding confusion) but I understand this can make things easier to read for reviewers. Besides, how can I fight with Ken Beck? 😅
So I pushed a commit with the change in all files I could think of (on the backend).

Reply:
[Verifiers] We should discuss it offline (and outside of this PR) as I'm not sure to get the point. I understand the introduction of deserializer and I agree this is a good step. But as far as signatures are concerned, I don't really get the improvement.
Can we agree this is out of scope?

Reply:
Yes this is out of scope. It is just that we have more time at this point of the year and I would like to frame better the refactorings related to the server-side. Even if it is not for this PR, it is for the long run.

Briefly, we need to instantiate an object of classes ourselves to start having business logic in it (including the implementation of the `Signed` interface). As far as I know, the following piece of code breaks:

```typescript
class MyRequest {
  value: boolean;
  method() {}
}

//...

const json = JSON.parse(str);
const request = json as MyRequest;
request.method() // method is not there
```
 

Thanks to that we can simplify a lot by replacing the verifiers and definitions full of generic with only one class.

Reply:
Yes, absolutely, a class needs to be instanciated.
As for the generics, it would be simpler to get rid of it, however it's still not clear to me how we can achieve this:
1. limitate duplication (building the "signature data", signing and verifying is similar for most messages)
2. have each message implement its own signature data.

Let's continue the discussion outside of this PR.